oauth/scope.js

```diff
@@ -0,0 +1,514 @@
+ * pair of cross-referenced maps, `this._values` and `this._implicants`,
+ * both of which are keyed by scope string values.
+ *
+ * The `_values` map has as its keys the actual scope values contained
+ * in the set, and the corresponding items are the set of implicants
+ * for each scope value.  Conversely, the `_implicants` map has as its
+ * keys every possible implicant of a value in the set, and the corresponding
+ * items are the set of scope values that are implied by that implicant.
+ *
+ * Thus:
+ *
+ *  - if `v in this._values` then the value `v` is a member of this
+ *    set, and `this._values[v]` gives all possible values that might
+ *    imply it.
+ *
+ *  - if `v in this._implicants` then the value `v` implies one or more
+ *    value in this set, and `this._implicants[v]` gives all the values
+ *    in this set that are implied by it.
+ *
+ * This scheme allows us to implement efficient membership checks, set
+ * comparison, and set union with deduplication of redundant values.
+ *
+ * The tradeoff here is memory usage, as we have to store the full
+ * set of implicants for each scope value.  We expect that this will not
+ * be too problematic in practice, because the size of this set is linear
+ * in the length of the scope value. A more memory-friendly implementation
+ * could use a trie-like data structure to efficiently store all implicants,
+ * with prefix compression.
+ *
+ *
+ */
+class Scope {
+
+  constructor(values, ...parentScopes) {
+    this._values = {};
+    this._implicants = {};
+    values.forEach(v => {
+      let implicants;
+      // If we're building a new `Scope` based on some existing parents,
+      // then we can probably find the implicants of each value as an
+      // existing set in one of those parents, rather than having to
+      // parse the value strings again.
+      for (const parentScope of parentScopes) {
+        implicants = parentScope._values[v];
+        if (implicants) {
+          break;
+        }
+      }
+      if (! implicants) {
+        implicants = new Set();
+        for (const implicant of getImplicantValues(v)) {
+          implicants.add(implicant);
+        }
+      }
+      this._addValue(v, implicants);
+    })
+  }
+
+  _addValue(value, implicants) {
+    // If this value is an implicant of some already-added value(s),
+    // then those values are redundant and can be removed.
+    if (value in this._implicants && ! (value in this._values)) {
+      for (const implied of this._implicants[value]) {
+        this._removeValue(implied);
+      }
+    }
+    // If this value is implied by some already-added value,
+    // it's redundant and we can ignore it.
+    let implied = false;
+    for (const implicant of implicants) {
+      if (implicant in this._values) {
+        implied = true;
+        break;
+      }
+    }
+    // Otherwise, add it to our set.
+    if (! implied) {
+      this._values[value] = implicants;
```

Comment:
Setting `this._values[value] = implicants` makes it seem like `this._values` is misnamed and it should instead be `this._implicants`, leaving me to wonder what `this._implicants` actually is.

Reply:
These have been renamed per your suggestion, and I think it's a lot clearer now.  Thanks!